src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -6355,6 +6408,416 @@ register_handler(int sig, struct sigaction *old_handler)
+			if (rc)
+				goto free_eq;
+			D_MUTEX_INIT(&aio_eq_list[idx + i].lock, NULL);
+			atomic_store_relaxed(&aio_eq_list[idx + i].n_op_queued, 0);
+			atomic_store_relaxed(&aio_eq_list[idx + i].n_op_done, 0);
+		}
+		aio_eq_count += num_aio_eq_create;
+	}
+	aio_ctx->first_eq = depth_sum_accu % aio_eq_count;
+	depth_sum_accu += aio_ctx->depth;
+	D_MUTEX_UNLOCK(&lock_aio);
+
+	if (aio_eq_count == 0) {
+		DS_ERROR(EBUSY, "no EQs created for AIO contexts");
+		return EBUSY;
+	}
+
+	D_MUTEX_LOCK(&aio_ctx->lock);
+	if (aio_ctx->inited == false) {
+		D_ALLOC_ARRAY(aio_ctx->ev_done_list, aio_ctx->depth);
+		if (aio_ctx->ev_done_list == NULL) {
+			D_MUTEX_UNLOCK(&aio_ctx->lock);
+			return ENOMEM;
+		}
+		/* empty queue */
+		aio_ctx->ev_done_h       = 0;
+		aio_ctx->ev_done_t       = 0;
+		aio_ctx->ev_done_list[0] = NULL;
+		aio_ctx->ev_queue_len    = 0;
+		aio_ctx->inited          = true;
+	}
+	aio_ctx->on_dfs = true;
+	D_MUTEX_UNLOCK(&aio_ctx->lock);
+
+	return 0;
+
+free_eq:
+	for (j = 0; j < i; j++) {
+		daos_eq_destroy(aio_eq_list[idx + j].eq, 0);
+	}
+	D_MUTEX_UNLOCK(&lock_aio);
+	return daos_der2errno(rc);
+}
+
+int
+io_submit(io_context_t ctx, long nr, struct iocb *ios[])
+{
+	aio_ctx_t     *aio_ctx_obj = (aio_ctx_t *)ctx;
+	io_context_t   ctx_real    = aio_ctx_obj->ctx;
+	daos_size_t    read_size   = 0;
+	d_iov_t        iov         = {};
+	d_sg_list_t    sgl         = {};
+	struct aio_ev *ctx_ev;
+	int            i, n_op_dfs, fd, io_depth;
+	int            idx_aio_eq;
+	int            rc;
+	short          op;
+	long int       aio_eq_count_local;
+
+	if (next_io_submit == NULL) {
+		next_io_submit = dlsym(RTLD_NEXT, "io_submit");
+		D_ASSERT(next_io_submit != NULL);
+	}
+	if (!hook_enabled)
+		return next_io_submit(ctx_real, nr, ios);
+	io_depth = aio_ctx_obj->depth;
+	if (io_depth == 0)
+		return next_io_submit(ctx_real, nr, ios);
+
+	n_op_dfs = 0;
+	for (i = 0; i < nr; i++) {
+		if (ios[i]->aio_fildes >= FD_FILE_BASE)
+			n_op_dfs++;
+	}
+	if (n_op_dfs == 0)
+		return next_io_submit(ctx_real, nr, ios);
+
+	if (n_op_dfs != nr) {
+		DS_ERROR(EINVAL, "io_submit() does not support mixed non-dfs and dfs files yet");
+		return (-EINVAL);
+	}
+
+	/* only support IO_CMD_PREAD and IO_CMD_PWRITE */
+	for (i = 0; i < nr; i++) {
+		op = ios[i]->aio_lio_opcode;
+		if (op != IO_CMD_PREAD && op != IO_CMD_PWRITE) {
+			DS_ERROR(EINVAL, "io_submit only supports PREAD and PWRITE for now");
+			return (-EINVAL);
+		}
+	}
+
+	aio_eq_count_local = aio_eq_count;
+	for (i = 0; i < nr; i++) {
+		op         = ios[i]->aio_lio_opcode;
+		fd         = ios[i]->aio_fildes - FD_FILE_BASE;
+		idx_aio_eq = (aio_ctx_obj->first_eq + i) % aio_eq_count_local;
+
+		D_ALLOC_PTR(ctx_ev);
+		if (ctx_ev == NULL) {
+			if (i == 0)
+				return -ENOMEM;
+			else
+				return i;
+		}
+
+		rc = daos_event_init(&ctx_ev->ev, aio_eq_list[idx_aio_eq].eq, NULL);
+		if (rc) {
+			D_FREE(ctx_ev);
+			DL_ERROR(rc, "daos_event_init() failed");
+			D_GOTO(err, rc = daos_der2errno(rc));
```

Comment:
so above you returned success / num io submitted on a failure, but here you disregard i and return error always on failure.. so there is inconsistency in behavior.

Reply:
You are right. Now they are handled in the same way. Thank you!